README.md
# msToString

## Library for converting a number of milliseconds into a time string such as "HH:MM:SS"

This library provides a C code function named msToString() that converts a long integer (uint32_t type) value that represents a number of milliseconds into a string that takes one of the forms "HH", "HH:MM", "HH:MM:SS", "MM", "MM:SS", "SS", or any of those followed by a decimal point and fractional digits. The string could be printed to a device such as the serial monitor port or a display, for example.

## Example usage

Here is sample code showing how to use the `msToString()` function. It also uses the `monitor_printf` library to printf to the serial monitor window.

```
#include <msToString.h>
#include <monitor_printf.h>

uint32_t MS;
char S[20];
const char n = sizeof(S);

void setup() {
  monitor.begin(&Serial, 115200);
  MS = 0;
  monitor.printf("Initial elapsed time is %s\n", msToString(MS, S, n));
}

void loop() {
  delay(1000);
  MS += 1000;
  monitor.printf("Current elapsed time is %s\n", msToString(MS, S, n));
}
```

The string output from msToString() above is of the form "HHH:MM:SS" giving the number of hours, minutes, and seconds that MS milliseconds equates to, ignoring extra milliseconds short of a full second.

Note that the colon character, ":", is currently fixed and cannot be changed to a different character.

## msToString() declaration

The `msToString()` function has this declaration:

```
char* msToString(uint32_t MS, char* S, size_t n, bool hours=true,
  bool minutes=true, bool seconds=true, int numDigits=0, int digitsAfterDP=0,
  bool* exceededMax=NULL);
```

The arguments are:

> **MS:** the long integer (uint32_t type) value that is the number of milliseconds to convert to a string.
>
> **S:** a character buffer in which to place the result, must be at least as long as the longest expected result.
>
> **n:** sizeof(S), the number of characters in the character buffer S. Result is truncated beyond this.
>
> **hours:** true to show hours in the result, false to omit hours, default is true.
>
> **minutes:** true to show minutes in the result, false to omit minutes, default is true.
>
> **seconds:** true to show seconds in the result, false to omit seconds, default is true.
>
> **numDigits:** >0 to use leading zeroes in the first number to make it have that many digits, 0 for no limit on the number of digits in it. The other two numbers always use leading zeroes so they have 2 digits, since they must be minutes or seconds and range from 0 to 59. The first number can be either hours, minutes, or seconds depending on the preceding arguments. The default for numDigits is 0. If >0 and the value of MS gives a first number with more digits, the entire result is shown with "X"'s instead of digits, to indicate overflow of result. For example, if all three numbers are shown and numDigits is 2, then if MS is less than 100 hours, the output string will be of the form ##:##:## with leading zeroes used when necessary to make 2 digits. However, if MS is >= 100 hours, then the output string will be "XX:XX:XX", with X's used to indicate overflow beyond the desired 2 digits.
>
> **digitsAfterDP:** number of digits to show after a decimal point at the end of the string, default is 0. If 0, no decimal point and fractional digits are shown. Otherwise, whatever number of milliseconds remain beyond what is shown by the hours, minutes, and/or seconds shown are displayed as a fraction of the last number shown. For example, if digitsAfterDP is 2 and hours and minutes are shown but seconds are not, then if 6000 milliseconds are left over after computing the hours and minutes numbers, then ".10" would appear after the "HH:MM" numbers, for example "2:45.10", since 6000 milliseconds is one-tenth of a minute.
>
> **exceededMax:** is normally NULL, which is its default value. If not NULL, it is a pointer to a bool variable, and on return from this function that variable is set *false* if the first number did not exceed the limit of digits given by numDigits, or *true* if it did exceed that. If numDigits=0 the variable will always be *false*.

The function returns the value of `S`, which allows a call to the function to be used as an argument to a function, such as for a printf %s-specifier's value, as in the example where monitor.printf() uses a "%s" specifier and its value is then provided by calling msToString().

## Example of HH:MM or MM:SS

As an example of using the *hours*, *minutes*, and *seconds* arguments, if you wanted to only show hours and minutes, you would use a call like this, with *seconds=false*:

```
  monitor.printf("Elapsed time in HH:MM is %s\n", msToString(MS, S, n, true, true, false));
```

Or, if you wanted to only show minutes and seconds, you would use a call like this, with *hours=false*:

```
  monitor.printf("Elapsed time in MM:SS is %s\n", msToString(MS, S, n, false, true, true));
```

It makes no sense to call for HH:SS form, so if *hours=seconds=true* and *minutes=false*, this is treated the same as if *minutes* were also true.

You can also set just one of *hours*, *minutes*, and *seconds* true and the other two false, to show only the number of hours, or minutes, or seconds, as a single integer value. For example:

```
  monitor.printf("Elapsed time in minutes is %s\n", msToString(MS, S, n, false, true, false));
```

## Example where first number is a fixed length

Often you want to force the first number of the result to be a fixed length, usually 2 characters, since clock time is usually shown using two digits. This is done by setting *numDigits* to the number of digits desired for the first number. For example:

```
  monitor.printf("Elapsed time in HH:MM:SS is %s\n", msToString(MS, S, n, true, true, true, 2));
```

If the first number has fewer digits, leading 0's are inserted to make it have exactly *numDigits* digits. For example, if MS equated to "1:45" in the previous example, it would instead print "01:45".

Setting numDigits to a number other than 0 also has the effect of forcing digits to be replaced with X's when the first number has *more than numDigits digits*. For example, if MS equated to "187:45" in the previous example, it would instead print "XX:XX".

## Including a decimal fraction in the result

Sometimes you may want to have a result string that includes a decimal point and fractional value. This is done by setting *digitsAfterDP* to the number of digits desired after the decimal point. The interpretation of the fraction as fractional hours, fractional minutes, or fractional seconds depends on what the last number before the decimal point is. For example:

```
  monitor.printf("Elapsed time in HH:MM.MMM is %s\n", msToString(MS, S, n, true, true, false, 2, 3));
```

In that example the last number before the decimal point is *minutes*, so the fraction part represents fractional minutes.

Note that rounding *is not done* on the final digit.

The decimal point character, ".", is currently fixed, it cannot be changed to a different character.

# Using false for hours, minutes, and seconds

You can also set all three of *hours*, *minutes*, and *seconds* to false. This causes conversion of MS to an integer string, ignoring *digitsAfterDP* but honoring *numDigits* (which can be used to insert leading 0's to pad the result to a specific length and replace digits with 'X's if there are more than that many digits). It is not expected that the user will use this, but you could. It is used internally by msToString() in recursive calls to itself to convert each of the hours, minutes, and seconds values.

## Contact

If you find a problem, please submit an issue report [here](https://github.com/tedtoal/msToString/issues/new/choose).

You can contact me (the author) at [ted@tedtoal.net](ted@tedtoal.net)


CHANGELOG.md
# msToString change log
---

## Table of Contents

* [Release 1.2.1](#release-121)
* [Release 1.1.1](#release-111)
---

### Release 1.2.1

1. Fix numbers after the first to ALWAY USE LEADING ZEROES to fill out the number to exactly two digits (for minutes and/or seconds following an initial hours or minutes). The *numDigitsFN* argument is renamed to *numDigits*.

### Release 1.1.1

1. Initial release.


src/msToString.cpp
/*
  msToString.cpp - A function to convert a number of milliseconds into a time
  string of the form (HH)(:MM)(:SS)(.##) in a buffer.
  Created by Ted Toal, July 19, 2023.
  Released into the public domain.


  Software License Agreement (BSD License)

  Copyright (c) 2023 Ted Toal
  All rights reserved.

  Redistribution and use in source and binary forms, with or without
  modification, are permitted provided that the following conditions are met:
  1. Redistributions of source code must retain the above copyright
  notice, this list of conditions and the following disclaimer.
  2. Redistributions in binary form must reproduce the above copyright
  notice, this list of conditions and the following disclaimer in the
  documentation and/or other materials provided with the distribution.
  3. Neither the name of the copyright holders nor the
  names of its contributors may be used to endorse or promote products
  derived from this software without specific prior written permission.

  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS ''AS IS'' AND ANY
  EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
  WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
  DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER BE LIABLE FOR ANY
  DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
  (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
  ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
  (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/
#include <Arduino.h>
#include <stdlib.h>
#include <stdio.h>
#include <msToString.h>
#
char* msToString(uint32_t MS, char* S, size_t n, bool hours, bool minutes,
  bool seconds, int numDigits, int digitsAfterDP, bool* exceededMax) {

  bool _exceededMax = false;
  size_t len;
  char* p = S;
  uint32_t div = 0;

  // hours=minutes=seconds=false
  // Plain conversion of MS to an integer string, honoring numDigits.
  if (!hours && !minutes && !seconds) {
    if (numDigits == 0)
      snprintf(p, n, "%02ld", MS);
    else {
      int Ndigits = 1;
      for (uint32_t V = 10; MS >= V && V != 1000000000; V *= 10)
        if (++Ndigits > numDigits) {
          _exceededMax = true;
          break;
        }
      }
      if (!_exceededMax) {
        char fmt[8];
        sprintf(fmt, "%%0%dld", numDigits); // e.g. "%02ld"
        snprintf(p, n, fmt, MS);
      } else {
        while (numDigits > 0 && n > 0) {
          *p++ = 'X';
          n--;
          numDigits--;
        }
        *p = 0;
    }
  }

  // hours=true
  if (hours) {
    div = 3600000UL; // That many milliseconds in an hour.
    uint16_t hours = (uint16_t) (MS / div);
    MS %= div;
    // Use recursive call with hours=minutes=seconds=false to convert hours to
    // a string, honoring numDigits.
    msToString(hours, p, n, false, false, false, numDigits, 0, &_exceededMax);
    numDigits = 2;
    len = strlen(p);
    p = p + len;
    n -= len;
  }

  // minutes=true OR hours=seconds=true
  if (minutes || (hours && seconds)) {
    if (p != S && n > 0) {
      *p++ = ':';
      *p = 0;
      n--;
    }
    if (_exceededMax) {
      strncpy(p, "XX", n);
      p[n] = 0;
    } else {
      div = 60000U; // That many milliseconds in a minute.
      uint32_t minutes = MS / div;
      MS %= div;
      msToString(minutes, p, n, false, false, false, numDigits, 0, &_exceededMax);
      numDigits = 2;
    }
    len = strlen(p);
    p = p + len;
    n -= len;
  }

  // seconds=true
  if (seconds) {
    if (p != S && n > 0) {
      *p++ = ':';
      *p = 0;
      n--;
    }
    if (_exceededMax) {
      strncpy(p, "XX", n);
      p[n] = 0;
    } else {
      div = 1000U; // That many milliseconds in a second.
      uint32_t seconds = MS / div;
      MS %= div;
      msToString(seconds, p, n, false, false, false, numDigits, 0, &_exceededMax);
      numDigits = 2;
    }
    len = strlen(p);
    p = p + len;
    n -= len;
  }

  // If requested, add decimal point and following digits.
  if (digitsAfterDP > 0 && div > 0 && n > 0) {
    *p++ = '.';
    *p = 0;
    n--;
    // We need to compute the first digitsAfterDP of MS/div.
    while (digitsAfterDP-- && n > 0) {
      if (_exceededMax) {
        *p++ = 'X';
        *p = 0;
      } else {
        MS *= 10;
        int digit = (int) (MS/div);
        MS %= div;
        sprintf(p++, "%d", digit);
      }
      n--;
    }
  }

  if (exceededMax != NULL)
    *exceededMax = _exceededMax;
  return(S);
}

// -------------------------------------------------------------------------


src/msToString.h
/*
  msToString.h - A function to convert a number of milliseconds into a time
  string of the form (HH)(:MM)(:SS)(.##) in a buffer.
  Created by Ted Toal, July 19, 2023.
  Released into the public domain.


  Software License Agreement (BSD License)

  Copyright (c) 2023 Ted Toal
  All rights reserved.

  Redistribution and use in source and binary forms, with or without
  modification, are permitted provided that the following conditions are met:
  1. Redistributions of source code must retain the above copyright
  notice, this list of conditions and the following disclaimer.
  2. Redistributions in binary form must reproduce the above copyright
  notice, this list of conditions and the following disclaimer in the
  documentation and/or other materials provided with the distribution.
  3. Neither the name of the copyright holders nor the
  names of its contributors may be used to endorse or promote products
  derived from this software without specific prior written permission.

  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS ''AS IS'' AND ANY
  EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
  WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
  DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER BE LIABLE FOR ANY
  DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
  (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
  ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
  (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/
#ifndef msToString_h
#define msToString_h

#include <Arduino.h>

/**************************************************************************/
/*!
  @brief    Convert a number of milliseconds into a string.
  @param    MS        A count of a number of milliseconds.
  @param    S         Result buffer of a size large enough to hold the result.
  @param    n         sizeof(S).
  @param    hours     true to include hours in string, false for not.
  @param    minutes   true to include minutes in string, false for not. Ignored
                      and taken as true if 'hours' and 'minutes' are both true.
  @param    seconds   true to include seconds in string, false for not.
  @param    numDigits Determines whether LEADING ZEROES are used for the first
                      number (hours, minutes, or seconds depending on previous
                      arguments) and the other numbers (minutes and/or seconds).
                      Possible values:
                        > 0: the first number has numDigits digits and the
                             others have 2 digits. E.g. 2 means all numbers have
                             exactly 2 digits, while 3 means the first number
                             has 3 digits and the others have 2 digits.
                        0:   leading zeroes are only used for numbers after the
                             first. The first number has any number of digits,
                             while the others have exactly 2 digits since they
                             range from 0 to 59.
                      For example, if all three numbers are shown and numDigits
                      is 2, then if MS is less than 100 hours, the output string
                      will be of the form ##:##:## with leading zeroes used when
                      necessary to make 2 digits. However, if MS is >= 100 hours
                      then the output string will be "XX:XX:XX", with X's used
                      to indicate overflow beyond the desired 2 digits.
  @param    digitsAfterDP   Number of digits to include after a decimal point
                            following the last number. If 0, no decimal point is
                            included. The final digit is NOT ROUNDED.
  @param    exceededMax     If not NULL, *exceededMax is set true if the number
                            of digits in the first number exceeded numDigits,
                            or is set false if not (or if numDigits=0).
  @returns                  S.
  @note     If the resulting string is longer than n-1 characters, the remaining
            characters are discarded and not stored in S.
  @note     If all of 'hours', 'minutes', 'seconds' are false, MS is converted
            into string representing number of milliseconds, and digitsAfterDP
            is ignored. numDigits is still honored. Thus 'milliseconds' is
            irrelevant, MS is simply an integer to be converted to a string.
  @note     The colon character, ":", is used to separate hours from minutes
            from seconds and is currently fixed, it cannot be changed to a
            different character.
  @note     The decimal point character, ".", used when digitsAfterDP > 0, is
            currently fixed, it cannot be changed to a different character.
*/
/**************************************************************************/
extern char* msToString(uint32_t MS, char* S, size_t n, bool hours=true,
  bool minutes=true, bool seconds=true, int numDigits=0, int digitsAfterDP=0,
  bool* exceededMax=NULL);

#endif // msToString_h
